Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -490,30 +530,66 @@ public CsharpCompiler(ErrorReporter reporter)
     /// </summary>
     private void CompileDatatypeDowncastClone(DatatypeDecl datatype, ConcreteSyntaxTree wr,
         List<TypeParameter> nonGhostTypeArgs, bool toInterface = false, bool lazy = false, DatatypeCtor ctor = null) {
-      if (nonGhostTypeArgs.Any(ty => ty.Variance == TypeParameter.TPVariance.Contra)
-          || datatype.Ctors.Any(ctor => ctor.Formals.Any(f => !f.IsGhost && (f.Type.IsArrowType || f.Type.IsRefType)))) {
-        return;
-      }
-      var typeArgs = TypeParameters(nonGhostTypeArgs, uniqueNames: true);
-      var dtArgs = "_I" + datatype.CompileName + typeArgs;
+      bool InvalidType(Type ty, bool refTy) =>
+        (ty.AsTypeParameter != null && refTy && datatype.TypeArgs.Contains(ty.AsTypeParameter))
+        || ty.TypeArgs.Exists(arg => InvalidType(arg, refTy || ty.IsRefType));
+      if (datatype.Ctors.Any(ctor => ctor.Formals.Any(f => !f.IsGhost && InvalidType(f.SyntacticType, false)))) { return; }
+      var customReceiver = nonGhostTypeArgs.Any(ty => ty.Variance == TypeParameter.TPVariance.Contra);
+      var uTypeArgs = TypeParameters(nonGhostTypeArgs, uniqueNames: true);
+      var typeArgs = TypeParameters(nonGhostTypeArgs);
+      var dtArgs = "_I" + datatype.CompileName + uTypeArgs;
+      var converters = $"{nonGhostTypeArgs.Comma((_, i) => $"converter{i}")}";
       string PrintConverter(TypeParameter tArg, int i) {
         var name = IdName(tArg);
         return $"Func<{name}, __{name}> converter{i}";
       }
 
       if (!toInterface) {
-        wr.Write($"public {((ctor != null || lazy) ? (datatype.IsRecordType ? "" : "override ") : "abstract ")}");
+        var modifiers = customReceiver
+          ? ((ctor != null || lazy) ? (datatype.IsRecordType ? "static " : "new static ") : "static ")
+          : ((ctor != null || lazy) ? (datatype.IsRecordType ? "" : "override ") : "abstract ");
+        wr.Write($"public {modifiers}");
       }
 
-      wr.Write($"{dtArgs} DowncastClone{typeArgs}({nonGhostTypeArgs.Comma(PrintConverter)})");
+      if (!(toInterface && customReceiver)) {
+        var receiver = customReceiver ? $"{"_I" + datatype.CompileName + typeArgs} _this, " : "";
+        wr.Write($"{dtArgs} DowncastClone{uTypeArgs}({receiver}{nonGhostTypeArgs.Comma(PrintConverter)})");
+      }
 
       if (ctor == null && !lazy) {
```

Comment:
It seems that there's two big blocks, one for `ctor == null && !lazy` and the other for not that. Can we name those blocks and put them in separate methods?

Reply:
What overhead are you referring to?

I'm suggesting something like:

```
if (ctor == null && !lazy) {
    ComeUpWithAName(datatype, wr, toInterface, customReceiver, uTypeArgs, typeArgs, converters, lazyClass);
    return;
  }

  ComeUpWithAnotherName(datatype, wr, nonGhostTypeArgs, lazy, ctor, customReceiver, uTypeArgs, typeArgs, dtArgs, converters, lazyClass);
 ```

Reply:
The one hidden in the enormous number of arguments. When you want to understand what any of those mean, you need to jump back to the caller.

Reply:
> When you want to understand what any of those mean, you need to jump back to the caller.

Why? Can't you name the arguments and the method so you don't need to jump back? I generally don't need to inspect the caller(s) of a method to understand what values the parameters of a method might contain.

Also, could you explain to me in words what these two cases are. Suppose you would create those two methods, what would the names be?

Also, as a guideline for method size, I like to keep it small enough so it fits on one screen.

Reply:
I guess that's where we differ. For me the truth is in the definition.

In one case we generate the actual `DowncastClone`, in the other only a signature. In some special circumstance the signature becomes a jump table.

Reply:
> I guess that's where we differ. For me the truth is in the definition.

Isn't the main point of naming things so you don't need to read their definition? For example, if I see a method body like this:

```
var program = Parse(input);
var resolvedProgram = Resolve(program);
var output = CodeGen(resolvedProgram);
```

Then I have some assumption on what the three called methods do, even if I don't dive into them, which will likely make it easier for me to navigate to whichever code I'm interested in reading more about. If the three methods had been inlined, then I would have many more lines to inspect before knowing where to go.

Reply:
Yes, that would be ideal, but generating such a name is not trivial in general.

Reply:
I fully agree with @keyboardDrummer's points here. Breaking things up into smaller methods has a lot of advantages: the cognitive load is lower because you have to hold less information in your head at once to understand it, but also compilers/runtimes tend to bias towards making many small methods fast as well. Having a correct implementation is great, but it's far better to have a correct implementation that can be efficiently maintained and expanded by future humans. The Dafny code base isn't necessarily a shining example everywhere as lots of older code didn't go through the review process we're now imposing, so we have some catching up to do.

However, in this exact instance I wouldn't insist on the refactoring in this PR. I think it's introducing scope creep, and since the poor organization was already there and isn't being introduced, it's going to be a judgement call about whether it needs doing NOW. And in some cases biasing for understanding the PR diff is more important.

I'd suggest a comment or two to document @keyboardDrummer's observation about the flow, possibly with a TODO.

Can I leave it to the two of you to follow this rough guidance for any other bones of contention in this PR? :)